### PR TITLE
Add bind modes tests and run on multi-numa machine

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
### Description
Continuation of https://github.com/oneapi-src/unified-memory-framework/pull/241.
- Rebased on the latest main
- Added multi_numa workflow

Ref: https://github.com/oneapi-src/unified-memory-framework/issues/242

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [ ] CI workflows execute properly

### CODE REVIEW

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 1]
```diff
@@ -75,14 +115,20 @@ struct testNumaNodes : public testing::TestWithParam<int> {
...
     ::testing::ValuesIn(get_available_numa_nodes_numbers()));
 
+INSTANTIATE_TEST_SUITE_P(testNumaNodesAllocationsAllCpus, testNumaOnAllCpus,
+                         ::testing::ValuesIn(get_available_cpus()));
+
 // Test for allocations on numa nodes. This test will be executed for all numa nodes
 // available on the system. The available nodes are returned in vector from the
 // get_available_numa_nodes_numbers() function and passed to test as parameters.
-TEST_P(testNumaNodes, checkNumaNodesAllocations) {
+TEST_P(testNumaOnAllNodes, checkNumaNodesAllocations) {
```
Comment (igchor) [inline test/provider_os_memory_multiple_numa_nodes.cpp:131]: I think it would be good to have a test case for BIND where you set multiple NUMA nodes in node mask. This is allowed according to the documentation.

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 2]
```diff
@@ -104,3 +150,269 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
...
+    // This pointer must point to an initialized value before retrieving a number of
+    // the numa node that the pointer was allocated on (calling get_mempolicy).
+    memset(ptr, 0xFF, alloc_size);
+    int retrieved_numa_node_number = getNumaNodeByPtr(ptr);
+    ASSERT_EQ(retrieved_numa_node_number, numa_node_number);
+}
+
+// Test for allocations on numa nodes with interleave mode enabled.
+// The page allocations are interleaved across the set of nodes specified in nodemask.
+TEST_F(testNuma, checkModeInterleave) {
```
Comment (igchor) [inline test/provider_os_memory_multiple_numa_nodes.cpp:289]: Please also add a test where nodemask does not contain all NUMA nodes (e.g. a test with just one node).

**Review by igchor — COMMENTED** [review]
Are you going to address the following in this PR?:

> We should check that policies like 'preferred' actually fall back to other nodes in case there is not enough memory on the specified node. Also, for modes that should fail if there is not enough memory we should test they actually do fail. One way to do that would be to add some kind of env variable which, if set, we could use to specify NUMA node on which we want to allocate (and use it for all binding tests). Then, we could run those tests in an environment with some custom NUMA nodes

File: .github/workflows/multi_numa.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    name: ${{matrix.os}}
+    # run only on upstream; forks will not have the HW
+    if: github.repository == 'oneapi-src/unified-memory-framework'
+
+    strategy:
+      matrix:
+        os: [ubuntu-22.04, RHEL-9]
+        build_type: [Release]
+        compiler: [{c: gcc, cxx: g++}]
+        shared_library: ['OFF']
```
Comment (PatKamin) [inline .github/workflows/multi_numa.yml:20]: I think that variables with only one value shouldn't be included in the matrix as they don't generate new job runs.

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 1 above)
Comment (igchor) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: I think we rely on the mbind behavior (proximity of the nodes) so it would be good to verify that.

Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: makes sense, you're right. Done.

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 4]
```diff
@@ -101,9 +149,342 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
...
+    int retrieved_numa_node_number = getNumaNodeByPtr(ptr);
+    ASSERT_EQ(retrieved_numa_node_number, numa_node_number);
+    CPU_FREE(mask);
+}
+
+// Test for allocation on numa node with default mode enabled.
+// Since no policy is set by the set_mempolicy function, it should
+// default to the system-wide default policy, which allocates pages
+// on the node of the CPU that triggers the allocation.
+TEST_F(testNuma, checkModeDefault) {
```
Comment (PatKamin) [inline test/provider_os_memory_multiple_numa_nodes.cpp:316]: This test could be also ran on all possible cpus.

**Review by PatKamin — APPROVED** [review]

File: .github/workflows/qemu.yml [hunk 5]
```diff
@@ -123,5 +123,4 @@ jobs:
         fi
 
         scp -P 2222 ${{github.workspace}}/scripts/qemu/run-build.sh cxltest@127.0.0.1:/home/cxltest
-        scp -P 2222 ${{github.workspace}}/.github/scripts/install_hwloc.sh cxltest@127.0.0.1:/home/cxltest
```
Comment (bratpiorka) [inline .github/workflows/qemu.yml:126]: is this change related to this PR?

**Review by bratpiorka — APPROVED** [review]

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 6]
```diff
@@ -101,9 +149,342 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
...
+TEST_F(testNuma, checkModeInterleave) {
+    constexpr int pages_num = 1024;
+    size_t page_size = sysconf(_SC_PAGE_SIZE);
+    umf_os_memory_provider_params_t os_memory_provider_params =
+        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+
+    std::vector<unsigned> numa_nodes = get_available_numa_nodes();
+    set_all_available_nodemask_bits(nodemask);
+
+    os_memory_provider_params.numa_list = &numa_nodes[0];
```
Comment (kswiecicki) [inline test/provider_os_memory_multiple_numa_nodes.cpp:348]: nit: `numa_nodes.data()` does the same trick.

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 7]
```diff
@@ -101,9 +149,342 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
...
+    memset(ptr, 0xFF, pages_num * page_size);
+
+    // Test where each page will be allocated.
+    // Get the first numa node for ptr; Each next page is expected to be on next nodes.
+    size_t index = getNumaNodeByPtr((char *)ptr);
+    for (size_t i = 1; i < (size_t)pages_num; i++) {
+        index++;
+        if (index == (size_t)numa_nodes.size()) {
+            index = 0;
+        }
```
Comment (kswiecicki) [inline test/provider_os_memory_multiple_numa_nodes.cpp:369]: nit: `index = (index + 1) % numa_nodes.size()`

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 4 above)
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp]: Done.

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 7 above)
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:382]: Done.

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 6 above)
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:361]: Done.

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 8]
```diff
@@ -101,9 +149,342 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
...
+    memset(ptr, 0xFF, alloc_size);
+    int retrieved_numa_node_number = getNumaNodeByPtr(ptr);
+    ASSERT_EQ(retrieved_numa_node_number, numa_node_number);
+}
+
+// Test for allocations on numa nodes with interleave mode enabled.
+// The page allocations are interleaved across the set of nodes specified in nodemask.
+TEST_F(testNuma, checkModeInterleave) {
+    constexpr int pages_num = 1024;
+    size_t page_size = sysconf(_SC_PAGE_SIZE);
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:354]: These tests are run on Linux for now, anyway

File: .github/workflows/multi_numa.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+      - name: Get information about platform
+        run: .github/scripts/get_system_info.sh
+
+      - name: Configure build
+        run: >
+          cmake
+          -B ${{github.workspace}}/build
+          -DCMAKE_BUILD_TYPE=Release
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply .github/workflows/multi_numa.yml:31]: Done.

File: src/provider/provider_os_memory.c [hunk 10]
```diff
@@ -196,8 +196,11 @@ static umf_result_t translate_params(umf_os_memory_provider_params_t *in_params,
     result = translate_numa_mode(in_params->numa_mode, emptyNodeset,
                                  &provider->numa_policy);
     if (result != UMF_RESULT_SUCCESS) {
-        LOG_ERR("incorrect NUMA mode: %u", in_params->numa_mode);
+        LOG_ERR("incorrect NUMA mode (%u) or wrong params",
+                in_params->numa_mode);
         return result;
+    } else {
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory.c:202]: Done.

File: scripts/qemu/run-build.sh [hunk 11]
```diff
@@ -10,11 +10,7 @@ repo=$1
 branch=$2
 
 echo password | sudo -Sk apt update
-echo password | sudo -Sk apt install -y git cmake gcc g++ numactl libnuma-dev libjemalloc-dev libtbb-dev
-
-# install packages required for building hwloc from source
-echo password | sudo -Sk apt install -y autoconf libtool
-source install_hwloc.sh
```
[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply scripts/qemu/run-build.sh:17]: pkg-config was missing 😉

File: test/provider_os_memory_multiple_numa_nodes.cpp (hunk 4 above)
Comment (lukaszstolarczuk) [reply test/provider_os_memory_multiple_numa_nodes.cpp:316]: hmm... or not done... it started to failing with this change. I reverted it and will debug it later.
